### PR TITLE
PLATY-486 Support for terminating EC2 instance when virus detected

### PR DESCRIPTION
**Pull request by tomasz-rosiek**

### CODE REVIEW

File: app/services/ScanUploadedFilesFlow.scala [hunk 1]
```diff
@@ -27,8 +27,10 @@ class ScanUploadedFilesFlow @Inject()(
...
   parser: MessageParser,
   scanningService: ScanningService,
-  scanningResultHandler: ScanningResultHandler)(implicit ec: ExecutionContext)
+  scanningResultHandler: ScanningResultHandler,
+  ec2InstanceTerminator: InstanceTerminator)(implicit ec: ExecutionContext)
     extends PollingJob {
+
   def run(): Future[Unit] = {
     val outcomes = for {
       messages        <- consumer.poll()
```
Comment (konradwudkowski) [inline app/services/ScanUploadedFilesFlow.scala:36]: How many massages can there be? In the next line you are instantiating a scan for all messages concurrently which may fail if the number of messages is sufficiently large

File: app/services/ScanningResultHandler.scala [hunk 2]
```diff
@@ -22,10 +22,14 @@ import model.S3ObjectLocation
 
 import scala.concurrent.{ExecutionContext, Future}
 
+sealed trait InstanceSafety
```
Comment (konradwudkowski) [inline app/services/ScanningResultHandler.scala]: I always add `extends Product with Serializable` to my sealed traits for reasons described here: https://underscore.io/blog/posts/2015/06/04/more-on-sealed.html

File: test/services/ScanUploadedFilesFlowSpec.scala [hunk 3]
```diff
@@ -40,35 +41,74 @@ class ScanUploadedFilesFlowSpec extends UnitSpec with Matchers with GivenWhenThe
...
+      verify(queueConsumer).poll()
 
       And("scanning result handler is called")
-      Mockito
-        .verify(scanningResultHandler)
-        .handleScanningResult(FileIsClean(S3ObjectLocation("bucket", "ID")))
+      verify(scanningResultHandler)
+        .handleScanningResult(FileIsInfected(s3object, "Virus name"))
 
       And("successfully processed messages are confirmed")
```
Comment (konradwudkowski) [inline test/services/ScanUploadedFilesFlowSpec.scala:105]: label And("successfully processed messages are confirmed") could be moved closer to where confirmation happens (3 lines below)

File: test/services/ScanUploadedFilesFlowSpec.scala [hunk 4]
```diff
@@ -77,108 +117,100 @@ class ScanUploadedFilesFlowSpec extends UnitSpec with Matchers with GivenWhenThe
...
+      val s3object3 = S3ObjectLocation("bucket", "ID3")
+
       val queueConsumer = mock[QueueConsumer]
-      Mockito.when(queueConsumer.poll()).thenReturn(List(validMessage1, invalidMessage, validMessage2))
-      Mockito
-        .when(queueConsumer.confirm(any()))
+      when(queueConsumer.poll()).thenReturn(List(validMessage1, invalidMessage, validMessage2))
+      when(queueConsumer.confirm(any()))
         .thenReturn(Future.successful(()))
         .thenReturn(Future.successful(()))
```
Comment (konradwudkowski) [inline test/services/ScanUploadedFilesFlowSpec.scala:127]: are both `.thenReturn(Future.successful(()))` required?

File: test/services/ScanUploadedFilesFlowSpec.scala [hunk 5]
```diff
@@ -40,35 +41,74 @@ class ScanUploadedFilesFlowSpec extends UnitSpec with Matchers with GivenWhenThe
...
-      Mockito.verify(queueConsumer).confirm(validMessage)
+      And("instance is terminated")
+      val inOrder = Mockito.inOrder(queueConsumer, instanceTerminator)
+
+      inOrder.verify(queueConsumer).confirm(validMessage)
+      inOrder.verify(instanceTerminator).terminate()
+
     }
 
     "get messages from the queue consumer, and perform scanning for valid messages and ignore invalid messages" in {
```
Comment (konradwudkowski) [inline test/services/ScanUploadedFilesFlowSpec.scala:114]: It's not easy to see from this test why `invalidMessage` is ignored

File: test/services/ScanUploadedFilesFlowSpec.scala [hunk 6]
```diff
@@ -77,108 +117,100 @@ class ScanUploadedFilesFlowSpec extends UnitSpec with Matchers with GivenWhenThe
...
       val validMessage2 = Message("ID2", "VALID-BODY", "RECEIPT-2")
       val validMessage3 = Message("ID3", "VALID-BODY", "RECEIPT-3")
 
       val queueConsumer = mock[QueueConsumer]
-      Mockito.when(queueConsumer.poll()).thenReturn(List(validMessage1, validMessage2, validMessage3))
-      Mockito
-        .when(queueConsumer.confirm(any()))
+      when(queueConsumer.poll()).thenReturn(List(validMessage1, validMessage2, validMessage3))
+      when(queueConsumer.confirm(any()))
         .thenReturn(Future.successful(()))
```
Comment (konradwudkowski) [inline test/services/ScanUploadedFilesFlowSpec.scala:179]: are both .thenReturn(Future.successful(())) required?

File: app/services/ScanUploadedFilesFlow.scala (hunk 1 above)
Comment (tomasz-rosiek) [reply app/services/ScanUploadedFilesFlow.scala:36]: It's 1 :) But I will explicitly add this to the code.

**Pull request merged**